### PR TITLE
feat: enhance attention() with scale, attn_bias, softcap, and is_causal

### LINKED ISSUE #4475: Enhance burn::tensor::module::attention() with scale, float mask, and softcap
## Context

`burn::tensor::module::attention()` currently computes `softmax(QK^T / sqrt(d)) * V` with an optional bool mask. This is the right abstraction for backend-dispatched SDPA (enabling flash attention on CubeCL), but it lacks a few features needed to fully cover common attention patterns.

In [burn-onnx](https://github.com/tracel-ai/burn-onnx), we use `burn::tensor::module::attention()` for the ONNX Attention operator when possible, but fall back to manual tensor ops for three missing features. Adding them to the burn API would let more models benefit from flash attention automatically.

### 1. Custom scale parameter

**Current**: hardcoded `1/sqrt(d_k)` where `d_k` is the last dimension of Q.
**Requested**: `scale: Option<f64>` parameter, defaulting to `1/sqrt(d_k)` when `None`.

This is standard in other frameworks (PyTorch's `F.scaled_dot_product_attention` has a `scale` parameter). Some models use non-standard scaling.

We currently work around this by pre-scaling Q (`Q' = scale * sqrt(d) * Q`) so that `Q'K^T / sqrt(d) = scale * QK^T`, but native support would be cleaner and work correctly with flash attention kernels that may optimize the scaling differently.

### 2. Float additive mask (attention bias)

**Current**: only `Option<Tensor<B, 4, Bool>>` (positions set to `-inf`).
**Requested**: also accept a float tensor that gets **added** to attention scores before softmax.

Float masks are used for:
- ALiBi (Attention with Linear Biases)
- Relative position biases (T5, Swin Transformer)
- ONNX Attention's `attn_mask` when provided as float type

PyTorch's SDPA supports both bool and float masks via the same `attn_mask` parameter.

### 3. Softcap

**Current**: not supported.
**Requested**: `softcap: Option<f64>` parameter. When provided, applies `softcap * tanh(scores / softcap)` after QK matmul (and mask addition) but before softmax.

Used in Gemma-2 and related models. See [Gemma-2 blog post](https://huggingface.co/blog/gemma2#soft-capping-and-attention-implementations).

## Proposed signature

```rust
pub fn attention<B: Backend>(
    query: Tensor<B, 4>,
    key: Tensor<B, 4>,
    value: Tensor<B, 4>,
    mask: Option<Tensor<B, 4, Bool>>,
    // new parameters:
    scale: Option<f64>,
    attn_bias: Option<Tensor<B, 4>>,
    softcap: Option<f64>,
) -> Tensor<B, 4>
```

Or alternatively, an `AttentionOptions` struct to keep the API clean:

```rust
pub struct AttentionOptions<B: Backend> {
    pub scale: Option<f64>,
    pub attn_bias: Option<Tensor<B, 4>>,
    pub softcap: Option<f64>,
}
```

## Priority

From burn-onnx's perspective:
1. **Float additive mask** - most impactful, covers ALiBi and many real-world models
2. **Custom scale** - we have a workaround but native support is cleaner
3. **Softcap** - niche (Gemma-2 family) but increasingly common

## References

- ONNX Attention spec (opset 23+): https://onnx.ai/onnx/operators/onnx__Attention.html
- PyTorch SDPA: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
- burn-onnx issue: https://github.com/tracel-ai/burn-onnx/issues/14

### PR DESCRIPTION
**Pull request by antimora**
## Summary

Enhances `burn::tensor::module::attention()` with four new capabilities, enabling models like ALiBi, T5, Swin Transformer, and Gemma-2 to use the backend-dispatched attention path instead of falling back to manual tensor ops.

- **`AttentionOptions` struct** with `scale: Option<f64>`, `softcap: Option<f64>`, and `is_causal: bool`
- **`attn_bias: Option<Tensor<B, 4>>`** parameter for float additive masks (ALiBi, relative position biases)
- CubeCL backend falls back to naive attention for `attn_bias`, `softcap`, custom `scale`, and `is_causal` (flash attention kernel returns incorrect results with `causal=true` on some Metal GPUs, tracked in #4484)
- Softcap applied before `-inf` masking to preserve masking semantics (`tanh(-inf)` would produce finite values)
- Softcap validated to be positive (panics on zero to prevent silent NaN)

### New API

```rust
pub fn attention<B: Backend>(
    query: Tensor<B, 4>,
    key: Tensor<B, 4>,
    value: Tensor<B, 4>,
    mask: Option<Tensor<B, 4, Bool>>,
    attn_bias: Option<Tensor<B, 4>>,
    options: AttentionOptions,
) -> Tensor<B, 4>
```

### Files changed

- `crates/burn-backend/src/backend/ops/modules/base.rs` - `AttentionOptions` struct, `ModuleOps` trait
- `crates/burn-backend/src/backend/ops/modules/attention.rs` - naive attention implementation
- `crates/burn-tensor/src/tensor/module.rs` - public API
- `crates/burn-cubecl/src/ops/module.rs` - CubeCL backend with fallback logic
- `crates/burn-backend-tests/tests/tensor/float/module/attention.rs` - 8 tests

## Test plan

- [x] `cargo test -p burn-backend-tests --test tensor --no-default-features --features metal,std --release -- attention` (16 tests: 8 float + 8 f16)
- [x] `cargo test -p burn-backend-tests --test fusion --no-default-features --features metal,std --release -- attention` (16 tests: 8 float + 8 f16)
- [x] `cargo test -p burn-backend-tests --test tensor -- attention` (8 tests, ndarray)

Closes #4475

### CODE REVIEW

File: crates/burn-backend/src/backend/ops/modules/attention.rs [hunk 1]
```diff
@@ -39,3 +74,31 @@ pub fn naive_attention<B: Backend>(
...
+        B::int_arange(0..seq_k as i64, &device),
+        Shape::new([1, seq_k]),
+    );
+
+    // mask where col > row + offset (upper triangle)
+    let rows_shifted = B::int_add_scalar(rows, offset.into());
+    let mask_2d = B::int_lower(rows_shifted, cols);
+
+    // Expand to [batch_size, num_heads, seq_q, seq_k]
+    B::bool_expand(mask_2d, Shape::new([batch_size, num_heads, seq_q, seq_k]))
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-backend/src/backend/ops/modules/attention.rs]: Good catch. While the 2D->4D expand works on both ndarray and CubeCL backends (they both handle rank changes), explicit reshape first is safer and more readable. Fixed: now reshapes to `[1, 1, seq_q, seq_k]` before expanding.

File: crates/burn-backend/src/backend/ops/modules/base.rs [hunk 2]
```diff
@@ -993,11 +1010,14 @@ pub trait ModuleOps<B: Backend> {
     /// optionally applying a mask to the attention scores.
     ///
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-backend/src/backend/ops/modules/base.rs]: Fixed. Updated the doc to describe the actual behavior: `softmax(QKᵗ * scale) · V` with scale defaulting to `1/sqrt(head_dim)`, plus the new options.

File: crates/burn-tensor/src/tensor/module.rs [hunk 3]
```diff
@@ -494,18 +494,21 @@ pub fn linear<B: Backend, const D: usize>(
     }
 }
 
-/// Computes scaled dot-product attention: softmax(QKᵗ / √d) · V,
-/// optionally applying a 4D mask to the attention scores.
+/// Computes scaled dot-product attention: softmax(QKᵗ * scale) · V,
+/// optionally applying a mask and/or additive bias to the attention scores.
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-tensor/src/tensor/module.rs]: Fixed. Updated the doc to clarify that scale defaults to `1/sqrt(head_dim)` and is configurable via `options.scale`.

File: crates/burn-backend/src/backend/ops/modules/attention.rs [hunk 4]
```diff
@@ -2,33 +2,68 @@ use core::f32;
...
+    } else {
+        attention_scores
+    };
+
+    // Softcap: softcap * tanh(scores / softcap)
+    let attention_scores = if let Some(softcap) = options.softcap {
+        let scaled = B::float_div_scalar(attention_scores, softcap.into());
+        let tanh = B::float_tanh(scaled);
+        B::float_mul_scalar(tanh, softcap.into())
+    } else {
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-backend/src/backend/ops/modules/attention.rs]: Good catch. Fixed: moved softcap before any `-inf` masking so `tanh` only sees finite logits. Added a regression test (`test_attention_softcap_preserves_causal_mask`) that verifies position 0 with causal masking outputs exactly `value[0]`, which would fail if softcap corrupted the `-inf` values into finite `-softcap`.
